common/client-libs/gateway-client/src/client.rs

```diff
@@ -438,8 +442,25 @@ impl GatewayClient {
         if self.shared_key.is_some() {
```

Comment:
I don't quite follow, why is `perform_initial_authentication` returning an Arc now?

Reply:
The `register_with_gateway` functions of the native and socks5 clients called directly on the `register` function. With this change, they'll go through `perform_initial_authentication`, which in turn will call `register` for them + do the credential stuff too (which is anyway needed). The Arc was created anyway in the native and socks5 clients, only now it's all in one place.
So, to answer your question, the Arc was used anyway, but I did a bit of refactoring to have the creation of that Arc + the code for the coconut credential all in just one place.